README.rst
##############
Wabbit Wappa
##############

.. image:: https://travis-ci.org/mokelly/wabbit_wappa.svg?branch=master
    :target: https://travis-ci.org/mokelly/wabbit_wappa

**Wabbit Wappa** is a full-featured Python wrapper for the lightning fast `Vowpal Wabbit <https://github.com/JohnLangford/vowpal_wabbit/wiki>`_ ("VW") 
machine learning utility.  Wabbit Wappa makes it easier to use VW's powerful features while abstracting away its idiosyncratic syntax and interface.

.. contents:: :local:

****************
Features
****************

* Complete Pythonic wrapper for the Vowpal Wabbit training and test syntax
* Online training and testing, with no need to restart VW or reload the trained model to go between them
* Save the trained model on the fly

****************
Getting Started
****************

If you're unfamiliar with Vowpal Wabbit, this documentation is no substitute for 
the `detailed tutorials <https://github.com/JohnLangford/vowpal_wabbit/wiki/Tutorial>`_
at the VW wiki.  You'll eventually need to read those to understand VW's advanced features.

Installation
===============

You have three installation options, depending on your comfort with compiling and installing the VW utility.

**If you already have Vowpal Wabbit installed**::

    pip install wabbit_wappa

**If you still need to install VW (currently version 7.7) and its dependencies**:

Start by cloning the WW repository::

    git clone https://github.com/mokelly/wabbit_wappa.git
    cd wabbit_wappa

Then run the included install script (which more or less follows the VW instructions)::

    scripts/vw-install.sh
    export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/lib;
    python setup.py install

(The "export" line should be added to your .profile if you don't want to run it every time you use Vowpal Wabbit.)

**If you want a virtual machine with everything all set up for you:**
    
*Windows users, this is your only option at present*

First install the virtual machine manager `Vagrant <http://www.vagrantup.com/>`_ along with your favorite virtualization system (such as `VirtualBox <https://www.virtualbox.org/>`_).
Then from the Wabbit Wappa source directory type::

    vagrant up

This will launch an Ubuntu VM and provision it with VW and WW, completely automatically!  Once that's all complete, just SSH to your new VM with::

    vagrant ssh
    
Testing
---------

Make sure everything is installed and configured correctly by running the tests::

    py.test

Usage Example
===============

Let's walk through an example of using Wabbit Wappa.  We will teach VW to recognize
capitalized characters.
(You can find the whole script at ``examples/capitalization_demo.py``.)

Start a default VW process in Logistic Regression mode::

    >>> from wabbit_wappa import *
    >>> vw = VW(loss_function='logistic')
    >>> print vw.command
    vw --save_resume --quiet --loss_function logistic --predictions /dev/stdout

Behind the scenes, Wabbit Wappa generates a command line including default parameters critical
to interaction with this wrapper.  VW is immediately run as a subprocess.

Now train the logistic model by sending 10 labeled examples to the VW learner::

    for i in range(10):
        label, features = get_example()  # Random example; see capitalization_demo.py
        vw.send_example(label, features=features)

From examples like these::

    Label -1: ['z', 'x', 'n', 'F', 'C', 'B', 'f', 'p', 'O'] is mostly lowercase
    Label -1: ['S', 'u', 'e', 'K', 'f', 'w', 'l', 'C', 'd'] is mostly lowercase
    Label -1: ['g', 'v', 'q', 'z', 'x', 'B', 'T', 'p', 'M'] is mostly lowercase
    Label 1: ['j', 'i', 'k', 'D', 'm', 'N', 'Q', 'Z', 'L'] is mostly uppercase
    Label 1: ['B', 'U', 'V', 'R', 'i', 'h', 'T', 'A', 'v'] is mostly uppercase
    Label 1: ['Y', 'u', 'R', 'K', 's', 'X', 'g', 'M', 'j'] is mostly uppercase
    Label -1: ['t', 'L', 'a', 'g', 'D', 'E', 'f', 'G', 'u'] is mostly lowercase
    Label 1: ['F', 'W', 'y', 'i', 'U', 'E', 'X', 'r', 'e'] is mostly uppercase
    Label -1: ['s', 'e', 'h', 'U', 'J', 'C', 'j', 'P', 'b'] is mostly lowercase
    Label 1: ['A', 'k', 'H', 'G', 'a', 'b', 'w', 'Q', 'V'] is mostly uppercase

VW begins to find the pattern: a +1 label if the capital letters outnumber the
lowercase, and -1 otherwise.

How well trained is our model?  Let's run 100 tests on new random examples::

    for i in range(num_tests):
        label, features = get_example()
        # Give the features to the model, witholding the label
        response = vw.get_prediction(features)
        prediction = response.prediction
        # Test whether the floating-point prediction is in the right direction
        if cmp(prediction, 0) == label:
            num_good_tests += 1

(For logistic regression, a ``prediction`` value greater than zero representa
a label of +1; that is why ``cmp(prediction, 0)`` is used.)

    >>> print "Correctly predicted", num_good_tests, "out of", num_tests
    Correctly predicted 60 out of 100

We can go on training, without restarting the process.  Let's train on 1,000 more examples::

    for i in range(1000):
        label, features = get_example()
        vw.send_example(label, features=features)

Now how good are our predictions?

    Correctly predicted 98 out of 100

We can save the model to disk at any point in the process::

    filename = 'capitalization.saved.model'
    vw.save_model(filename)

and reload our model using the 'i' argument::

    >>> vw2 = VW(loss_function='logistic', i=filename)
    >>> print vw2.command
    vw -i capitalization.saved.model --save_resume --quiet --loss_function logistic --predictions /dev/stdout

The ``vw2`` model will now give just the same predictions that ``vw`` would have; and the default ``save_resume=True`` parameter
means we can continue training from where we left off.

To shut down the VW subprocess before your program exits, call ``vw.close()``.


****************
Documentation
****************

Namespaces
===============

The most important Vowpal Wabbit feature not discussed above is namespaces.  VW
uses namespaces to divide features into groups, which is used for some of its
advanced features.  Without discussing in detail *why* you would use them,
here's *how* to use namespaces in Wabbit Wappa.

To reproduce an example from this `Vowpal Wabbit tutorial <https://github.com/JohnLangford/vowpal_wabbit/wiki/v6.1_tutorial.pdf>`_::

    namespace1 = Namespace('excuses', 0.1, [('the', 0.01), 'dog', 'ate', 'my', 'homework'])
    namespace2 = Namespace('teacher', features='male white Bagnell AI ate breakfast'.split())

These namespaces can then be used as examples in training and prediction::

    vw.send_example(response=1.,
                    importance=.5,
                    tag="example_39",
                    namespaces=[namespace1, namespace2])
    response = vw.get_prediction(namespaces=[namespace1, namespace2])
    prediction = response.prediction

Alternatively, Namespaces can be queued up to be used automatically in the next
example or prediction sent to the VW subprocess::

    vw.add_namespace(namespace1)
    vw.add_namespace(namespace2)
    vw.send_example(response=-1., importance=.5, tag="example_39")

or::

    vw.add_namespace('excuses', 0.1, [('the', 0.01), 'dog', 'ate', 'my', 'homework'])
    vw.add_namespace('teacher', features='male white Bagnell AI ate breakfast'.split())
    response = vw.get_prediction()
    prediction = response.prediction

Tokens in Vowpal Wabbit may not contain the space character, ``:`` or ``|``.  By default,
Wabbit Wappa will detect and escape these characters::

    >>> namespace = Namespace('Metric Features', 3.28, [('hei|ght', 1.5), ('len:gth', 2.0)])
    >>> print namespace.to_string()
    Metric\_Features:3.28 hei\\ght:1.5 len\;gth:2.0

If you wish, you can get the raw VW input lines and pass them to the subprocess directly::

    vw.add_namespace(namespace1)
    vw.add_namespace(namespace2)
    raw_line = vw.make_line(response=1., importance=.5, tag="example_39")
    vw.send_line(raw_line)

    >>> print raw_line
    1.0 0.5 'example_39|excuses:0.1 the:0.01 dog ate my homework |teacher male white Bagnell AI ate breakfast


VW Options
===============

In the ``VW()`` constructor, each named argument corresponds
to a Vowpal Wabbit option.  Single character keys are mapped to single-dash options;
e.g. ``b=20`` yields ``-b 20``.  Multiple character keys map to double-dash options:
``quiet=True`` yields ``--quiet``.

Boolean values are interpreted as flags: present if True, absent if False (or not given).
All non-boolean values are treated as option arguments, as in the `-b` example above.

If an option argument is a list, that option is repeated multiple times;
e.g. ``q=['ab', 'bc']`` yields ``-q ab -q bc``.

Run ``vw -h`` from your terminal for a listing of most options.

Note that Wabbit Wappa makes no attempt to validate the inputs or
ensure they are compatible with its functionality.  For instance, changing the
default ``predictions='/dev/stdout'`` will probably make that ``VW()`` instance
non-functional.

Active Learning
=================

Active Learning is an approach to training somewhere between supervised and unsupervised.
When getting labeled data is very expensive (such as when users must be solicited for
their preferences), an Active Learning approach assigns an "importance" value to each
unlabeled example, so that only the most critical labels need be acquired.

Vowpal Wabbit's `Active Learning <https://github.com/JohnLangford/vowpal_wabbit/wiki/active_learning.pdf>`_
interface requires you to start a VW instance in server mode and communicate with it
via a socket.  Wabbit Wappa abstracts all that away, providing the same interface for both
regular and Active learning::

    vw = VW(loss_function='logistic', active_mode=True, active_mellowness=0.1)
    response = vw.get_prediction(features)
    if response.importance >= 1.:
        label = get_expensive_label(features)
        vw.send_example(label, features=features)

See ``examples/active_learning_demo.py`` for a fully worked example.


API Documentation
===================

For complete explanation of all parameters, refer to the docstrings::

    import wabbit_wappa
    help(wabbit_wappa)



HISTORY.rst
.. :changelog:

****************
History
****************

0.0.1 (2014-03-31)
=====================

* First release on GitHub

0.0.2 (2014-04-07)
=====================

* Good unit test coverage
* Full documentation, instructions and demo
* Added command line builder with Pythonic interface

0.1.2 (2014-04-09)
=====================

* Now installable using pip
* Updated VW version to 7.6
* Tweaked line detection to speed up process communication

0.2.0 (2014-04-13)
=====================

* Active Learning interface, with documentation and example script
* Minor performance boosts
* **Backwards-incompatible change:** ``get_prediction()`` now returns a ``VWResult`` object, with the predicted value accessible as ``result.prediction``.
  
0.3.0 (2014-10-16)
======================

* Python 3 compatibility (thanks `Antti Haapala <https://github.com/ztane>`_!)
* Much faster line buffering (50% overall speed improvement) (thanks `Antti Haapala <https://github.com/ztane>`_!)
* Updated VW version to 7.7
* Updated Ubuntu in VM to Trusty
* Travis integration


test/test_wabbit_wappa.py

import random
import os
import time

from wabbit_wappa import *


def test_namespace():
    namespace = Namespace('MetricFeatures', 3.28, [('height', 1.5), ('length', 2.0), 'apple', '1948'])
    namespace_string = namespace.to_string()
    assert namespace_string == 'MetricFeatures:3.28 height:1.5 length:2.0 apple 1948 '

    namespace = Namespace(None, 3.28, ['height', 'length'])
    namespace_string = namespace.to_string()
    assert namespace_string == ' height length '


def test_validation():
    try:
        namespace = Namespace('Metric Features', 3.28, [('height|', 1.5), ('len:gth', 2.0)],
                              escape=False)
    except WabbitInvalidCharacter:
        pass  # This is the correct behavior
    else:
        assert False, "to_string() should error out for these inputs when escape==False"


def test_escaping():
    namespace = Namespace('Metric Features', 3.28, [('height|', 1.5), ('len:gth', 2.0)])
    namespace_string = namespace.to_string()
    assert 'Metric Features' not in namespace_string
    assert '|' not in namespace_string
    assert 'len:gth' not in namespace_string


def test_command():
    command = make_command_line(predictions='/dev/stdout',
                                quiet=True,
                                save_resume=True,
                                compressed=True,
                                q_colon=['a', 'b'],
                                b=20,
                                )
    # Test that command has all expected elements
    assert 'vw ' in command
    assert '--predictions /dev/stdout' in command
    assert '--quiet' in command
    assert '--save_resume' in command
    assert '--compressed' in command
    assert '--q: a' in command
    assert '--q: a' in command
    assert '-b 20' in command
    assert '--b 20' not in command
    # Test that VW runs with this command
    vw = VW(command)


def test_training():
    # TODO: pytest probably has a framework for testing hyperparameters like this
    for active_mode in [False, True]:
        vw = VW(loss_function='logistic', active_mode=active_mode)
        # Train with an easy case
        for i in range(20):
            # Positive example
            vw.send_example(response=1.,
                            importance=2.,
                            tag='positive',
                            features=[('a', 1 + random.random()),
                                      ('b', -1 - random.random())]
                            )
            vw.send_example(response=-1.,
                            importance=.5,
                            tag='negative',
                            features=[('lungfish', 1 + random.random()),
                                      ('palooka', -1 - random.random())]
                            )
        prediction1 = vw.get_prediction([('a', 1),
                                        ('b', -2)]).prediction
        # Prediction should be definitively positive
        assert prediction1 > 1.
        prediction2 = vw.get_prediction([('lungfish', 3)]).prediction
        # Prediction should be negative
        assert prediction2 < 0
        prediction3 = vw.get_prediction([('a', 1),
                                        ('b', -2)]).prediction
        # Making predictions shouldn't affect the trained model
        assert prediction1 == prediction3

        # Continue training with very different examples
        for i in range(20):
            # Positive example
            vw.add_namespace('space1',
                             1.0,
                             ['X', 'Y', 'Z'],
                             )
            vw.send_example(response=1.)
            # Negative example
            vw.add_namespace('space2',
                             2.0,
                             ['X', 'Y', 'Z'],
                             )
            vw.send_example(response=-1.)
        vw.add_namespace('space1',
                         1.0,
                         ['X'],
                         )
        prediction4 = vw.get_prediction().prediction
        # Prediction should be positive
        assert prediction4 > 0
        vw.add_namespace('space2',
                         1.0,
                         ['X'],
                         )
        prediction5 = vw.get_prediction().prediction
        # Prediction should be negative
        assert prediction5 < 0

        # Save the model to a temporary file
        filename = '__temp.model'
        vw.save_model(filename)
        # This sleep is required only in active_mode, in the (unusual) case
        # that the model file is used immediately
        time.sleep(0.1)

        # Load a new VW instance from that model
        vw2 = VW(loss_function='logistic', i=filename)
        # Make the same prediction with each model (testing cache_string to boot)
        namespace1 = Namespace(features=[('a', 1), ('b', -2)], cache_string=True)
        namespace2 = Namespace('space1', 1.0, ['X', 'Y'], cache_string=True)
        prediction1 = vw.get_prediction(namespaces=[namespace1, namespace2]).prediction
        prediction2 = vw2.get_prediction(namespaces=[namespace1, namespace2]).prediction
        assert prediction1 == prediction2
        assert prediction1 > 1.

        # Clean up
        vw.close()
        vw2.close()
        os.remove(filename)


wabbit_wappa/__init__.py
# -*- coding: utf-8 -*-
from __future__ import print_function, division, absolute_import, unicode_literals

"""
Wrapper for Vowpal Wabbit executable

TODO: 
-Detect VW version in unit tests; for command line generation scenarios,
    unit tests should detect whether it works as expected.
-Scenario assistance; e.g. caching examples for reuse in multi-pass
    (This would make good example code also.)
    -Abstraction for passes (with automatic usage of example cache)
-Example for README: Active learning interface
-Sklearn compatibility (like vowpal_porpoise)


by Michael J.T. O'Kelly, 2014-2-24
"""

__author__ = "Michael J.T. O'Kelly"
__email__ = 'mokelly@gmail.com'
__version__ = '0.3.0'

try:
    basestring
except:
    basestring = str

import logging
import re

import pexpect

from . import active_learner

class WabbitInvalidCharacter(ValueError):
    pass


validation_regex = re.compile(r' |:|\|')

def validate_vw_string(s):
    """Throw a WabbitInvalidCharacter exception if the string is
    not a 
    (http://stats.stackexchange.com/questions/28877/finding-the-best-features-in-interaction-models)
    """
    if validation_regex.search(s):
        raise WabbitInvalidCharacter(s)


escape_dict = {' ': r'\_',
               ':': r'\;',
               '|': r'\\'
               }

def escape_vw_character(special_character_re_match):
    special_character = special_character_re_match.group()
    return escape_dict[special_character]


def escape_vw_string(s):
    escaped_s = validation_regex.sub(escape_vw_character, s)
    return escaped_s


class Namespace():
    """Abstraction of Namespace part of VW example lines"""
    def __init__(self,
                 name=None,
                 scale=None,
                 features=None,
                 escape=True,
                 validate=True,
                 cache_string=False):
        """Create a namespace with given (optional) name and importance,
        initialized with any given features (described in add_features()).
        If 'validate', name and features are validated for compatibility
            with VW's reserved characters, throwing a WabbitInvalidCharacter
            exception.
        If 'escape', any invalid characters are replaced with escape characters.
            ('escape' mode supersedes 'vaildate' mode.)
        If 'cache_string', the results of any to_string() call are cached
            permanently, ignoring any further changes to self.  (This can
            speed things up if this Namespace is re-used.)
        """
        self.name = name
        self.scale = scale
        self.validate = validate
        self.escape = escape
        self._string = None
        self.features = []
        self.cache_string = cache_string
        if name:
            if escape:
                self.name = escape_vw_string(self.name)
            elif validate:
                validate_vw_string(self.name)
        if features:
            self.add_features(features)

    def add_features(self, features):
        """Add features to this namespace.
        features: An iterable of features.  A feature may be either
            1) A VW label (not containing characters from escape_dict.keys(),
                unless 'escape' mode is on)
            2) A tuple (label, value) where value is any float
        """
        for feature in features:
            if isinstance(feature, basestring):
                label = feature
                value = None
            else:
                label, value = feature
            self.add_feature(label, value)

    def add_feature(self, label, value=None):
        """
        label: A VW label (not containing characters from escape_dict.keys(),
            unless 'escape' mode is on)
        value: float giving the weight or magnitude of this feature
        """
        if self.escape:
            label = escape_vw_string(label)
        elif self.validate:
            validate_vw_string(label)
        feature = (label, value)
        self.features.append(feature)

    def to_string(self):
        """Export this namespace to a string suitable for incorporation
        in a VW example line, e.g.
        'MetricFeatures:3.28 height:1.5 length:2.0 '
        """
        if self._string is None:
            tokens = []
            if self.name:
                if self.scale:
                    token = self.name + ':' + str(self.scale)
                else:
                    token = self.name
            else:
                token = ''  # Spacing element to indicate next string is a feature
            tokens.append(token)
            for label, value in self.features:
                if value is None:
                    token = label
                else:
                    token = label + ':' + str(value)
                tokens.append(token)
            tokens.append('')  # Spacing element to separate from next pipe character
            output = ' '.join(tokens)
            if self.cache_string:
                self._string = output
        else:
            output = self._string
        return output

    def export_features(self, delimiter='\\'):
        """Export the features from this namespace as a list of feature tokens
        with the namespace name prepended (delimited by 'delimiter').

        Example:
        >>> namespace = Namespace('name1', features=['feature1', 'feature2'])
        >>> print namespace.export_features()
        ['name1\\feature1', 'name1\\feature2']
        """
        result_list = []
        for feature in self.features:
            result = '{}{}{}'.format(self.name, delimiter, feature)
            result_list.append(result)
        return result_list


class VWResult():
    """Parses VW string output into consistent structure"""
    def __init__(self, result_string, active_mode=False):
        """Set 'active_mode' to True to parse results in
        an Active Learning context."""
        self.raw_output = result_string
        result_list = []
        # TODO: Something more robust than whitespace splitting
        #   to handle modes like --audit ?
        for token in result_string.split():
            try:
                result = float(token)
                result_list.append(result)
            except ValueError:
                # Ignore tokens that can't be made into floats (like tags)
                logging.debug("Ignoring non-float token {}".format(token))
        self.value_list = result_list
        if result_list:
            self.prediction = result_list[0]
        else:
            self.prediction = None
        if active_mode:
            if len(result_list) > 1:
                self.importance = result_list[1]
            else:
                self.importance = 0.

    def __str__(self):
        return str(self.__dict__)


class VW():
    """Wrapper for VW executable, handling online input and outputs."""
    def __init__(self, command=None, active_mode=False, dummy_mode=False, **kwargs):
        """'command' is the full command-line necessary to run VW.  E.g.
        vw --loss_function logistic -p /dev/stdout --quiet
        -p /dev/stdout --quiet is mandatory for compatibility,
        and certain options like 
            --save_resume
        are suggested, while some options make no sense in this context:
            -d
            --passes
        wabbit_wappa.py does not support any mode that turns off piping to
        stdin or stdout

        active_mode: Launch VW in port-listening active learning mode, controlled via
            a simulated subprocess.
        dummy_mode: Don't actually start any VW process.  (Used for assembling
            VW command lines separately.)

        If no command is given, any additional keyword arguments are passed to
            make_command_line() and the resulting command is used.  (This provides
            sensible defaults.)
        """
        if command is None:
            if active_mode:
                active_settings = active_learner.get_active_default_settings()
                # Overwrite active settings with kwargs
                active_settings.update(kwargs)
                kwargs = active_settings
                port = kwargs.get('port')
            command = make_command_line(**kwargs)
        self.active_mode = active_mode
        self.dummy_mode = dummy_mode
        if dummy_mode:
            self.vw_process = None
        else:
            if active_mode:
                self.vw_process = active_learner.ActiveVWProcess(command, port=port)
            else:
                self.vw_process = pexpect.spawn(command)
                # Turn off delaybeforesend; this is necessary only in non-applicable cases
                self.vw_process.delaybeforesend = 0
                self.vw_process.setecho(False)
        logging.info("Started VW({})".format(command))
        self.command = command
        self.namespaces = []
        self._line = None

    def send_line(self, line, parse_result=True):
        """Submit a raw line of text to the VW instance, returning a 
        VWResult() object.

        If 'parse_result' is False, ignore the result and return None.
        """
        self.vw_process.sendline(line)  # Send line, along with newline
        result = self._get_response(parse_result=parse_result)
        return result

    def _get_response(self, parse_result=True):
        """If 'parse_result' is False, ignore the received output and return None."""
        # expect_exact is faster than just exact, and fine for our purpose
        # (http://pexpect.readthedocs.org/en/latest/api/pexpect.html#pexpect.spawn.expect_exact)
        # searchwindowsize and other attributes may also affect efficiency
        self.vw_process.expect_exact('\r\n', searchwindowsize=-1)  # Wait until process outputs a complete line
        if parse_result:
            output = self.vw_process.before
            result_struct = VWResult(output, active_mode=self.active_mode)
        else:
            result_struct = None
        return result_struct

    def send_example(self,
                     *args,
                     **kwargs
                     ):
        """Send a labeled or unlabeled example to the VW instance.
        If 'parse_result' kwarg is False, ignore the result and return None.

        All other parameters are passed to self.send_line().

        Returns a VWResult object.
        """
        # Pop out the keyword argument 'parse_result' if given
        parse_result = kwargs.pop('parse_result', True)
        line = self.make_line(*args, **kwargs)
        result = self.send_line(line, parse_result=parse_result)
        return result

    def make_line(self,
                  response=None,
                  importance=None,
                  base=None,
                  tag=None,
                  features=None,
                  namespaces=None,
                  ):
        """Makes and returns an example string in VW syntax.
        If given, 'response', 'importance', 'base', and 'tag' are used
        to label the example.  Features for the example come from
        any given features or namespaces, as well as any previously
        added namespaces (using them up in the process).
        """
        if namespaces is not None:
            self.add_namespaces(namespaces)
        if features is not None:
            namespace = Namespace(features=features)
            self.add_namespace(namespace)
        substrings = []
        tokens = []
        if response is not None:
            token = str(response)
            tokens.append(token)
            if importance is not None:  # Check only if response is given
                token = str(importance)
                tokens.append(token)
                if base is not None:  # Check only if importance is given
                    token = str(base)
                    tokens.append(token)
        if tag is not None: 
            token = "'" + str(tag)  # Tags are unambiguous if given a ' prefix
            tokens.append(token)
        else:
            token = ""  # Spacing element to avoid ambiguity in parsing
            tokens.append(token)
        substring = ' '.join(tokens)
        substrings.append(substring)
        if self.namespaces:
            for namespace in self.namespaces:
                substring = namespace.to_string()
                substrings.append(substring)
        else:
            substrings.append('')  # For correct syntax
        line = '|'.join(substrings)
        self._line = line
        self.namespaces = []  # Reset namespaces after their use
        return line

    def add_namespace(self, *args, **kwargs):
        """Accepts two calling patterns:
        add_namespace(namespace): queue a preexisting namespace onto
            this VW instance.
        add_namespace(name, scale, features, ...): Pass all args and kwargs
            to the Namespace constructor to make a new Namespace instance,
            and queue it to this VW instance.

        Returns self (so that this command can be chained).
        """
        if args and isinstance(args[0], Namespace):
            namespace = args[0]
        elif isinstance(kwargs.get('namespace'), Namespace):
            namespace = kwargs.get('namespace')
        else:
            namespace = Namespace(*args, **kwargs)
        self.namespaces.append(namespace)
        return self

    def add_namespaces(self, namespaces):
        """Add these namespaces sequentially.
        Returns self (so that this command can be chained)."""
        for namespace in namespaces:
            self.add_namespace(namespace)
        return self

    def get_prediction(self, features=None, tag=None, namespaces=None):
        """Send an unlabeled example to the trained VW instance.
        Uses any given features or namespaces, as well as any previously
        added namespaces (using them up in the process).

        Returns a VWResult object."""
        if features is not None:
            namespace = Namespace(features=features)
            self.add_namespace(namespace)
        result = self.send_example(tag=tag, namespaces=namespaces)
        return result

    def save_model(self, model_filename):
        """Pass a "command example" to the VW subprocess requesting
        that the current model be serialized to model_filename immediately.
        """
        line = "save_{}|".format(model_filename)
        self.vw_process.sendline(line)
        # No response is expected in this case

    def close(self):
        """Shut down the VW process."""
        self.vw_process.close()
        # TODO: Give this a context manager interface

    # TODO: Fancy interface for auditing data?


def make_command_line(predictions='/dev/stdout',
                      quiet=True,
                      save_resume=True,
                      q_colon=None,
                      **kwargs
                      ):
    """Construct a command line for VW, with each named argument corresponding
    to a VW option.
    Single character keys are mapped to single-dash options,
    e.g. 'b=20' yields '-b 20',
    while multiple character keys map to double-dash options:
        'quiet=True' yields '--quiet'
    Boolean values are interpreted as flags: present if True, absent if False.
    All other values are treated as option arguments, as in the -b example above.
    If an option argument is a list, that option is repeated multiple times,
    e.g. 'q=['ab', 'bc']' yields '-q ab -q bc'

    q_colon is handled specially, mapping to '--q:'.

    Run 'vw -h' for a listing of most options.

    Defaults are well-suited for use with Wabbit Wappa:
    vw --predictions /dev/stdout --quiet --save_resume

    NOTE: This function makes no attempt to validate the inputs or
    ensure they are compatible with Wabbit Wappa.

    Outputs a command line string.
    """
    args = ['vw']
    if q_colon:
        kwargs['q:'] = q_colon
    kwargs['predictions'] = predictions
    kwargs['quiet'] = quiet
    kwargs['save_resume'] = save_resume
    for key, value in kwargs.items():
        if len(key)==1:
            option = '-{}'.format(key)
        else:
            option = '--{}'.format(key)
        if value is True:
            arg_list = [option]
        elif isinstance(value, basestring):
            arg_list = ['{} {}'.format(option, value)]
        elif hasattr(value, '__getitem__'):  # Listlike value
            arg_list = [ '{} {}'.format(option, subvalue) for subvalue in value ]
        else:
            arg_list = ['{} {}'.format(option, value)]
        args.extend(arg_list)
    command = ' '.join(args)
    return command





